### ISSUE TITLE
About Multi-Backend Implementation of Gradient Checkpointing question

### ISSUE THREAD

**Issue by pass-lin** [keras-team/keras#19003]
I tried to implement Multi-Backend Gradient Checkpointing in https://github.com/pass-lin/bert4keras3
But I encounter some problems, such as when I implement in the tf backend
Class ScaleOffset (Layer):
    "Simple affine transformation layer (the last dimension is multiplied by the gamma vector and the beta vector is added)
    Description: 1. The specific operation is to multiply the gamma vector by the last dimension and add the beta vector;
         2. If scale and offset are specified directly, then direct constant scaling and translation;
         3, hidden_ * series parameters are only used when conditional input (conditional = True),
            It is used to control beta and gamma through external conditions.
    ""
    Def __init__ (
        Self,
        Scale = True,
        Offset = True,
        Conditional = False,
        hidden_units = None,
        hidden_activation = 'linear',
        hidden_initializer = 'glorot_uniform',
        ** kwargs
    ):
        Super (ScaleOffset, self). __init__ (** kwargs)
        Self.scale = scale
        Self.offset = offset
        Self.conditional = conditional
        Self. hidden_units = hidden_units
        Self.hidden_activation = activations.get (hidden_activation)
        Self.hidden_initializer = initializers.get (hidden_initializer)

    @integerize_shape
    Def build (self, input_shape):
        Super (ScaleOffset, self).build (input_shape)
        
        If self is conditional:
            input_shape = input_shape [0]

        If self.offset is True:
            Self.beta = self.add_weight (
                Name = 'beta', shape = (input_shape [-1],), initializer = 'zeros'
            )
        If self.scale is True:
            Self.gamma = self.add_weight (
                Name = 'gamma', shape = (input_shape [-1],), initializer = 'ones'
            )

        If self is conditional:

            If self. hidden_units is not None:
                Self. hidden_dense = Dense (
                    Units = self. hidden_units,
                    Activation = self. hidden_activation,
                    use_bias = False,
                    kernel_initializer = self hidden_initializer
                )

            If self.offset is not False and self.offset is not None:
                Self. beta_dense = Dense (
                    Units = input_shape [-1],
                    use_bias = False,
                    kernel_initializer = 'zeros'
                )
            If self.scale is not False and self.scale is not None:
                Self. gamma_dense = Dense (
                    Units = input_shape [-1],
                    use_bias = False,
                    kernel_initializer = 'zeros'
                )

    Def compute_mask (self, inputs, mask = None):
        If self is conditional:
            Return mask if mask is None else mask [0]
        Else:
            Return mask

    @Tf.recompute_grad
    Def call (self, inputs):
        "If there is a condition, the default input is list, and the second is the condition."
        ""
        If self is conditional:
            Inputs, conds = inputs
            If self. hidden_units is not None:
                Conds = self.hidden_dense (conds)
            Conds = align (conds, [0, -1], ops.ndim (inputs))

        If self.scale is not False and self.scale is not None:
            Gamma = self.gamma if self.scale is True else self.scale
            If self is conditional:
                Gamma = gamma + self. gamma_dense (conds)
            Inputs = inputs * gamma

        If self.offset is not False and self.offset is not None:
            Beta = self.beta if self.offset is True else self.offset
            If self is conditional:
                Beta = beta + self. beta_dense (conds)
            Inputs = inputs + beta

        Return inputs
I get an error when I type a keras_tensor name = 'beta', shape = (input_shape [-1],), initializer = 'zeros'

TypeError: 'NoneType' object is not subscriptable
Similarly, similar problems will be encountered in torch and jax. What is going on here, or can the official provide a multi-backend implementation?

**Comment by sachinprasadhs** [comment]
Hi, 

Could you please provide the simple standalone code to reproduce the issue in a colab Gist?

**Comment by pass-lin** [comment]
> Hi,
> 
> Could you please provide the simple standalone code to reproduce the issue in a colab Gist?

This is a simple, reproducible error code
`
#pip install bert4keras3==0.0.5
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
os.environ["KERAS_BACKEND"] = "tensorflow"
os.environ["CUDA_VISIBLE_DEVICES"]="-1"
os.environ['RECOMPUTE']="1"

from bert4keras3.models import build_transformer_model
bert =build_transformer_model(  attention_probs_dropout_prob= 0.0,
  directionality= "bidi",
  hidden_act= "gelu",
  hidden_size= 312,
  embedding_size= 128,
  initializer_range= 0.02,
  intermediate_size= 1248,
  max_position= 512,
  num_attention_heads= 12,
  num_hidden_layers= 4,
  type_vocab_size= 2,
  vocab_size= 21128,
  return_keras_model=True,)  # 建立模型，加载权重
`

**Comment by sachinprasadhs** [comment]
I think the code is still incomplete, please find the attached Gist [here](https://colab.sandbox.google.com/gist/sachinprasadhs/7066a0ea584c810d4f8ed4846764f633/19003.ipynb) with the error.

**Comment by pass-lin** [comment]
th

> 我认为代码仍然不完整，请[在此处](https://colab.sandbox.google.com/gist/sachinprasadhs/7066a0ea584c810d4f8ed4846764f633/19003.ipynb)找到带有错误的附件 Gist。

The error you are encountering is "TypeError: 'NoneType' object is not subscriptable." If you comment out `os.environ['RECOMPUTE']="1"`, you will find that the model can be loaded successfully.
`    def recompute_grad(call):
#The code in question is decorating the `call` function of the `dense` class.
        if os.environ['RECOMPUTE']=="1":
            return call
        return tf.recompute_grad(call)`

**Comment by pass-lin** [comment]
> Hi @pass-lin -
> 
> Could you please confirm if this issue is resolved for you ? Please feel free to close the issue if it is resolved!

yes,i have a easily implement https://github.com/keras-team/keras/pull/20720
Unfortunately, it seems to conflict with others' implementations.

**Comment by pctablet505** [comment]
fixed
